src/backend/distributed/executor/executor.c

```diff
@@ -2066,23 +2254,51 @@ PlacementExecutionDone(TaskPlacementExecution *placementExecution, bool succeede
 	}
 	else
 	{
+		if (ShouldMarkPlacementsInvalidOnFailure(execution))
+		{
+			ShardPlacement *shardPlacement = placementExecution->shardPlacement;
+
+			/*
+			 * We only set shard state if its current state is FILE_FINALIZED, which
+			 * prevents overwriting shard state if it is already set at somewhere else.
+			 */
+			if (shardPlacement->shardState == FILE_FINALIZED)
+			{
+				UpdateShardPlacementState(shardPlacement->placementId, FILE_INACTIVE);
+			}
+		}
+
 		placementExecution->executionState = PLACEMENT_EXECUTION_FAILED;
 	}
 
-	shardCommandExecution = placementExecution->shardCommandExecution;
 
-	/* determine the outcome of the overall task */
-	executionState = GetTaskExecutionState(shardCommandExecution);
-	if (executionState == TASK_EXECUTION_FINISHED)
+	/*
+	 * Update unfinishedTaskCount only when state changes from not finished to
+	 * finished or failed state.
+	 */
+	if (executionState == TASK_EXECUTION_NOT_FINISHED)
 	{
-		execution->unfinishedTaskCount--;
-		return;
-	}
+		TaskExecutionState newExecutionState =
+			TaskExecutionStateMachine(shardCommandExecution);
+		if (newExecutionState == TASK_EXECUTION_FINISHED)
+		{
+			execution->unfinishedTaskCount--;
+			return;
+		}
 
-	if (executionState == TASK_EXECUTION_FAILED)
+		if (newExecutionState == TASK_EXECUTION_FAILED)
+		{
+			execution->unfinishedTaskCount--;
+			execution->failed = true;
+			return;
+		}
+	}
+	else
 	{
-		execution->unfinishedTaskCount--;
-		execution->failed = true;
+		/*
+		 * Task execution has already been finished, no need to continue the
+		 * next placement.
+		 */
 		return;
```

Comment:
Wondering if this code could be structured a little more nicely, e.g..

```c
if (executionState != TASK_EXECUTION_NOT_FINISHED)
{
   /*
    * Task execution has already been finished, no need to continue the
    * next placement.
    */
    return;
}

newExecutionState = TaskExecutionStateMachine(shardCommandExecution);
if (newExecutionState == TASK_EXECUTION_FINISHED)
{
    // should this be in TaskExecutionStateMachine maybe?
    execution->unfinishedTaskCount--;
    return;
}
else if (newExecutionState == TASK_EXECUTION_FAILED)
{
    // should this be in TaskExecutionStateMachine maybe?
    execution->unfinishedTaskCount--;
    execution->failed = true;
    return;
}
else
{
    DoTheExecutionOrderThingWeDoBelow();
}
```

Reply:
makes sense